client_plugin/drivers/shared/etcdops.go

```diff
@@ -0,0 +1,269 @@
+	nodes, err := cli.NodeList(ctx, types.NodeListOptions{})
+	if err != nil {
+		log.WithFields(log.Fields{"nodeID": nodeID,
+			"error": err}).Error("Failed get NodeList from swarm manager")
+		return err
+	}
+	for _, n := range nodes {
+		if n.ManagerStatus != nil && n.ManagerStatus.Leader == true {
+			log.WithFields(
+				log.Fields{"leader ID": n.ID,
+					"manager ID": nodeID},
+			).Info("Found leader for swarm manager")
+
+			joinEtcdCluster(addr, n.ManagerStatus.Addr, nodeID)
+			return nil
+		}
+	}
+
+	err = fmt.Errorf("Failed to get leader for swarm manager %s", nodeID)
+	return err
+}
+
+// startEtcdCluster function is called by swarm leader to start a ETCD cluster
+func startEtcdCluster(nodeAddr string, nodeID string) error {
+	lines := []string{
+		"--name", nodeID,
+		"--advertise-client-urls", etcdScheme + nodeAddr + etcdClientPort,
+		"--initial-advertise-peer-urls", etcdScheme + nodeAddr + etcdPeerPort,
+		"--listen-client-urls", etcdScheme + etcdListenURL + etcdClientPort,
+		"--listen-peer-urls", etcdScheme + etcdListenURL + etcdPeerPort,
+		"--initial-cluster-token", etcdClusterToken,
+		"--initial-cluster", nodeID + "=" + etcdScheme + nodeAddr + etcdPeerPort,
+		"--initial-cluster-state", etcdClusterStateNew,
+	}
+	go etcdService(lines)
+
+	return nil
+}
+
+// joinEtcdCluster function is called by a non-leader swarm manager to join a ETCD cluster
+func joinEtcdCluster(nodeAddr string, leaderAddr string, nodeID string) error {
+	cli, err := addrToEtcdClient(leaderAddr)
+	if err != nil {
+		log.WithFields(
+			log.Fields{"nodeAddr": nodeAddr,
+				"leaderAddr": leaderAddr,
+				"nodeID":     nodeID},
+		).Error("Failed to join ETCD cluster on manager ")
+	}
+
+	// list all current ETCD members, check if this node is already added as a member
+	lresp, err := cli.MemberList(context.Background())
+	if err != nil {
+		log.WithFields(
+			log.Fields{"leaderAddr": leaderAddr,
+				"error":       err,
+				"members len": len(lresp.Members)},
+		).Error("Failed to list member for ETCD")
+		return err
+	}
+
+	peerAddr := etcdScheme + nodeAddr + etcdPeerPort
+	existing := false
+	for _, member := range lresp.Members {
+		if member.PeerURLs[0] == peerAddr {
+			if member.Name == "" {
+				// already added, not started
+				log.WithFields(
+					log.Fields{"nodeID": nodeID,
+						"peerAddr": peerAddr},
+				).Info("This node is already added as a etcd member, but not started yet ")
+
+				existing = true
+			} else {
+				// already added and started, need to reset
+				log.WithFields(
+					log.Fields{"nodeID": nodeID,
+						"peerAddr": peerAddr},
+				).Info("This node is already added as a etcd member and started, need to remove first ")
+
+				_, err = cli.MemberRemove(context.Background(), member.ID)
+				if err != nil {
+					log.WithFields(
+						log.Fields{"peerAddr": peerAddr,
+							"member.ID": member.ID},
+					).Error("Failed to remove member for ETCD")
+					return err
+				}
+			}
+			break
```

Comment:
Because we don't need to continue the comparison since one IP address can only join at once. No need to continue the check.
I will add some more comments here.